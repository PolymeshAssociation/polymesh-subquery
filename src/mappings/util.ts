import { decodeAddress } from '@polkadot/keyring';
import { Codec } from '@polkadot/types/types';
import { hexStripPrefix, u8aToHex, u8aToString } from '@polkadot/util';
import { SubstrateEvent, SubstrateExtrinsic } from '@subql/types';
import { Portfolio } from 'polymesh-subql/types/models/Portfolio';
import {
  AssetDocument,
  Compliance,
  Distribution,
  FoundType,
  SecurityIdentifier,
  TransferComplianceExemption,
  TransferManager,
  TransferRestrictionTypeEnum,
} from '../types';
import { Attributes } from './entities/common';

/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str[0].toLowerCase() + str.slice(1).replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);

export const snakeToCamelCase = (value: string): string =>
  value
    .toLowerCase()
    .replace(/([-_][a-z])/g, group => group.toUpperCase().replace('-', '').replace('_', ''));

export const capitalizeFirstLetter = (str: string): string => str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s?.replace(/\0/g, '') || '';

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (text: string, exitOnFirst = false): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case '(':
      case '{':
      case '<': {
        nestedLevel++;
        break;
      }
      case ')':
      case '}':
      case '>': {
        nestedLevel--;
        break;
      }
      case ',': {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(`No top level comma found in ${text}, it probably isn't a map`);
  }
  return commas;
};

export const getOrDefault = <K, V>(map: Map<K, V>, key: K, getDefault: () => V): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};

export const serializeAccount = (item: Codec): string | null => {
  const s = item.toString();

  if (s.trim().length === 0) {
    return null;
  }
  return u8aToHex(decodeAddress(item.toString(), false, item.registry.chainSS58));
};

export const getFirstKeyFromJson = (item: Codec): string => {
  return Object.keys(item.toJSON())[0];
};

export const getFirstValueFromJson = (item: Codec): string => {
  return item.toJSON()[getFirstKeyFromJson(item)];
};

export const getTextValue = (item: Codec): string => {
  return item?.toString().trim().length === 0 ? null : item.toString().trim();
};

export const getNumberValue = (item: Codec): number => {
  return Number(getTextValue(item));
};

export const getDateValue = (item: Codec): Date => {
  return item.toString().trim().length === 0 ? null : new Date(Number(item.toString()));
};

export const getBigIntValue = (item: Codec): bigint => {
  return BigInt(getTextValue(item) || 0);
};

export const getBooleanValue = (item: Codec): boolean => {
  return JSON.parse(getTextValue(item));
};

export const hexToString = (input: string): string => {
  const hex = hexStripPrefix(input);
  let str = '';
  for (let i = 0; i < hex.length; i += 2)
    str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
  return removeNullChars(str);
};

export const getSigner = (extrinsic: SubstrateExtrinsic): string => {
  return hexStripPrefix(
    u8aToHex(
      decodeAddress(
        extrinsic.extrinsic.signer.toString(),
        false,
        extrinsic.extrinsic.registry.chainSS58
      )
    )
  );
};

/**
 * Parses a raw Asset Document
 */
export const getDocValue = (
  doc: Codec
): Pick<AssetDocument, 'name' | 'link' | 'contentHash' | 'type' | 'filedAt'> => {
  const {
    uri: link,
    content_hash: documentHash,
    name,
    doc_type: type,
    filing_date: filedAt,
  } = JSON.parse(doc.toString());

  const hashType = Object.keys(documentHash)[0];

  return {
    name,
    link,
    contentHash: documentHash[hashType],
    type,
    filedAt,
  };
};

export const getSecurityIdentifiers = (item: Codec): SecurityIdentifier[] => {
  const identifiers = JSON.parse(item.toString());
  return identifiers.map(i => {
    const type = Object.keys(i)[0];
    return {
      type,
      value: i[type],
    };
  });
};

/**
 * Parses a Vec<AssetCompliance>
 */
export const getComplianceRulesValue = (
  requirements: Codec
): Pick<Compliance, 'complianceId' | 'data'>[] => {
  return JSON.parse(JSON.stringify(requirements));
};

/**
 * Parses AssetCompliance
 */
export const getComplianceValue = (
  compliance: Codec
): Pick<Compliance, 'complianceId' | 'data'> => {
  const { id, ...data } = JSON.parse(compliance.toString());
  return {
    complianceId: Number(id),
    data,
  };
};

/**
 * Parses AssetTransferManager
 */
export const getTransferManagerValue = (
  manager: Codec
): Pick<TransferManager, 'type' | 'value'> => {
  const { countTransferManager, percentageTransferManager } = JSON.parse(JSON.stringify(manager));

  if (countTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Count,
      value: Number(countTransferManager),
    };
  }

  if (percentageTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Percentage,
      value: Number(percentageTransferManager),
    };
  }

  throw new Error('Unknown transfer restriction type found');
};

export const getExemptKeyValue = (
  item: Codec
): Omit<Attributes<TransferComplianceExemption>, 'exemptedEntityId'> => {
  const {
    asset: { ticker },
    op: opType,
    claimType,
  } = JSON.parse(item.toString());
  return {
    assetId: hexToString(ticker),
    opType,
    claimType,
  };
};

export const getExemptionsValue = (exemptions: Codec): string[] => {
  return exemptions.toJSON() as string[];
};

export const logFoundType = (type: string, rawType: string): void => {
  FoundType.create({ id: type, rawType }).save();
};

export const END_OF_TIME = BigInt('253402194600000');

export function addIfNotIncludes<T>(arr: T[], item: T): void {
  if (!arr.includes(item)) {
    arr.push(item);
  }
}

interface MeshPortfolio {
  did: string;
  kind: {
    user: number;
  };
}

const meshPortfolioToPortfolio = ({
  did,
  kind: { user: number },
}: MeshPortfolio): Pick<Portfolio, 'identityId' | 'number'> => ({
  identityId: did,
  number: number || 0,
});

export const getPortfolioValue = (item: Codec): Pick<Portfolio, 'identityId' | 'number'> => {
  const meshPortfolio = JSON.parse(item.toString());
  return meshPortfolioToPortfolio(meshPortfolio);
};

export const getCaIdValue = (item: Codec): Pick<Distribution, 'localId' | 'assetId'> => {
  const { local_id: localId, ticker } = JSON.parse(item.toString());
  return { localId, assetId: hexToString(ticker) };
};

export interface LegDetails {
  from: Pick<Portfolio, 'identityId' | 'number'>;
  to: Pick<Portfolio, 'identityId' | 'number'>;
  ticker: string;
  amount: bigint;
}

export const getLegsValue = (item: Codec): LegDetails[] => {
  const legs = JSON.parse(item.toString());
  return legs.map(({ from: fromPortfolio, to: toPortfolio, asset: ticker, amount }) => ({
    from: meshPortfolioToPortfolio(fromPortfolio),
    to: meshPortfolioToPortfolio(toPortfolio),
    ticker: hexToString(ticker),
    amount: getBigIntValue(amount),
  }));
};

export const getSignerAddress = (event: SubstrateEvent): string => {
  let signer: string;
  if (event.extrinsic) {
    signer = getSigner(event.extrinsic);
  }
  return signer;
};

export const getDistributionValue = (
  item: Codec
): Pick<
  Distribution,
  'portfolioId' | 'currency' | 'perShare' | 'amount' | 'remaining' | 'paymentAt' | 'expiresAt'
> => {
  const { from, currency, per_share, amount, remaining, payment_at, expires_at } = JSON.parse(
    item.toString()
  );
  const { identityId, number } = meshPortfolioToPortfolio(from);
  return {
    portfolioId: `${identityId}/${number}`,
    currency: hexToString(currency),
    perShare: getBigIntValue(per_share),
    amount: getBigIntValue(amount),
    remaining: getBigIntValue(remaining),
    paymentAt: getBigIntValue(payment_at),
    expiresAt: getBigIntValue(expires_at || END_OF_TIME),
  };
};

export const logError = (message: string): void => {
  logger.error(message);
};
