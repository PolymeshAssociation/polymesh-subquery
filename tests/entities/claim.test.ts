import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('claimScopes', () => {
  it('should return an empty list as given identity has no claims', async () => {
    const identityId = '0x0200000000000000000000000000000000000000000000000000000000000001';
    const res = await query({
      query: gql`
        query {
          claimScopes(
            filter: {
              target: {
                equalTo: "${identityId}"
              }
            }
          ) {
            nodes {
              scope
              ticker
            }
          }
        }
      `,
    });

    return expect(res?.data?.claimScopes?.nodes?.length).toEqual(0);
  });

  it('should return a list of scopes for given identity', async () => {
    const identityId = '0x69650eb2544ed57930cc0bedacdfceeee3b5905470e56edb0eb96271e0e9fef3';
    const q = {
      query: gql`
      query {
        claimScopes(
          filter: {
            target: {
              equalTo: "${identityId}"
            }
          }
        ) {
          nodes {
            scope
            ticker
          }
        }
      }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });
});

describe('claims', () => {
  it('should return no registries as given identity has no claims', async () => {
    const identityId = '0x9900000000000000000000000000000000000000000000000000000000000000';
    const res = await query({
      query: gql`
        query{
          claims(filter: {
              targetId: {
                in: ["${identityId}"]
              }
            }) {
            totalCount
          }
        }
      `,
    });

    return expect(res?.data?.claims?.totalCount).toEqual(0);
  });

  it('should return just total count', async () => {
    const res = await query({
      query: gql`
        query {
          claims {
            totalCount
          }
        }
      `,
    });

    expect(res?.data).toMatchSnapshot();
  });

  it('should return all claims as there are no filters', async () => {
    const q = {
      query: gql`
        query {
          claims(first: 10) {
            totalCount
            nodes {
              targetId
              issuerId
              issuanceDate
              lastUpdateDate
              expiry
              filterExpiry
              type
              jurisdiction
              scope
              cddId
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });

  it('should return filtered claims', async () => {
    const scopeType = 'Identity';
    const scopeValue = '5TICKER';
    const trustedClaimIssuer = '0x56a91c10f2368b30670b7bea260928f0291387abfb75b5953cd722917423bf01';
    const q = {
      variables: { scopeValue },
      query: gql`
        query q($scopeValue: String!){
          claims(filter: {
            scope: {
              equalTo: {
                type: ${scopeType}, 
                value: $scopeValue
              }
            },
            issuerId: {
              equalTo: "${trustedClaimIssuer}"
            }
          }) {
            totalCount
            nodes {
              targetId
              issuerId
              issuanceDate
              lastUpdateDate
              expiry
              filterExpiry
              type
              jurisdiction
              scope
              cddId
            }
          }
        }
      `,
    };
    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });
});
