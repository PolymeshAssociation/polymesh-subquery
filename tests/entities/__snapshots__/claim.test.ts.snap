// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`claimScopes should return a list of scopes for given identity 1`] = `
Object {
  "claimScopes": Object {
    "__typename": "ClaimScopesConnection",
    "nodes": Array [
      Object {
        "__typename": "ClaimScope",
        "scope": Object {
          "type": "Identity",
          "value": "0x56a91c10f2368b30670b7bea260928f0291387abfb75b5953cd722917423bf01",
        },
        "ticker": null,
      },
      Object {
        "__typename": "ClaimScope",
        "scope": Object {
          "type": "Ticker",
          "value": "5TICKER",
        },
        "ticker": "5TICKER",
      },
    ],
  },
}
`;

exports[`claims should return all claims as there are no filters 1`] = `
Object {
  "claims": Object {
    "__typename": "ClaimsConnection",
    "nodes": Array [
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523618500",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523618500",
        "scope": null,
        "targetId": "0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000002",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523677000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523677000",
        "scope": null,
        "targetId": "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523570500",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523570500",
        "scope": null,
        "targetId": "0x139dad9675f6d89bbe39d4a1ffcb824b9ff44fce366390090752a926d7c6e689",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523663000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523663000",
        "scope": null,
        "targetId": "0x1c18365371d85ae8662de63bfb0fb90da99b93482e8f043de61ffc83994f8858",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523574000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523574000",
        "scope": null,
        "targetId": "0x25ebc60487a6e8c621b5987d0efa148a61be92f4810690b9a7d45298145f4ae8",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523660000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523660000",
        "scope": null,
        "targetId": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523662000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523662000",
        "scope": null,
        "targetId": "0x2fbb6ef977181d14c7714e134cd5382d42e4214aaf81a2e60b57663946b644c5",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": null,
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523888000",
        "issuerId": "0x0400000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523888000",
        "scope": Object {
          "type": "Ticker",
          "value": "15TICKER",
        },
        "targetId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "type": "Accredited",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523720500",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523720500",
        "scope": null,
        "targetId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "type": "CustomerDueDiligence",
      },
      Object {
        "__typename": "Claim",
        "cddId": "0x0000000000000000000000000000000000000000000000000000000000000002",
        "expiry": null,
        "filterExpiry": "253402194600000",
        "issuanceDate": "1633523560000",
        "issuerId": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "jurisdiction": null,
        "lastUpdateDate": "1633523560000",
        "scope": null,
        "targetId": "0x406c0c0af84c690b0fd730c53f9518af7b6a11f0aa5b1e3bdf48b5bf7f695fb4",
        "type": "CustomerDueDiligence",
      },
    ],
    "totalCount": 42,
  },
}
`;

exports[`claims should return filtered claims 1`] = `
Object {
  "claims": Object {
    "__typename": "ClaimsConnection",
    "nodes": Array [],
    "totalCount": 0,
  },
}
`;

exports[`claims should return just total count 1`] = `
Object {
  "claims": Object {
    "__typename": "ClaimsConnection",
    "totalCount": 42,
  },
}
`;
